src/main/kotlin/com/exactpro/th2/codec/http/HttpPipelineCodec.kt

```diff
@@ -129,37 +132,138 @@ class HttpPipelineCodec : IPipelineCodec {
+                }
+
+                RawHttpResponse(
+                    null,
+                    null,
+                    StatusLine(HTTP_1_1, statusCode, reason),
+                    headers.build(),
+                    null
+                ).run {
+                    body?.toBody().run(::withBody) ?: this
+                }
+            }
+
+            else -> error("Unsupported message type: $messageType")
         }
 
+        encodedMessages += RawMessage(
+            message.id,
+            message.eventId,
+            message.metadata,
+            message.protocol,
+            Unpooled.wrappedBuffer(httpMessage.toByteArray())
+        )
+
+        return MessageGroup(encodedMessages)
+    }
+
+    override fun decode(messageGroup: ProtoMessageGroup, context: IReportingContext): ProtoMessageGroup {
+        val messages = messageGroup.messagesList
+
         require(messages.size == 1) { "Message group must contain only 1 message" }
         require(messages[0].kindCase == RAW_MESSAGE) { "Message must be a raw message" }
 
         val message = messages[0].rawMessage
         val body = message.body.toByteArray().toString(UTF_8)
-        val builder = MessageGroup.newBuilder()
+        val builder = ProtoMessageGroup.newBuilder()
 
         when (val direction = message.metadata.id.direction) {
             FIRST -> RAW_HTTP.parseResponse(body).convert(RESPONSE_MESSAGE, message, builder) { httpMessage, _ ->
                 httpMessage.addField(STATUS_CODE_FIELD, statusCode)
                 httpMessage.addField(REASON_FIELD, reason)
             }
-            SECOND -> RAW_HTTP.parseRequest(body).convert(REQUEST_MESSAGE, message, builder) { httpMessage, metadataProperties ->
-                httpMessage.addField(METHOD_FIELD, method)
-                httpMessage.addField(URI_FIELD, uri)
-                metadataProperties[METHOD_METADATA_PROPERTY] = method
-                metadataProperties[URI_METADATA_PROPERTY] = uri.toString()
-            }
+
+            SECOND -> RAW_HTTP.parseRequest(body)
+                .convert(REQUEST_MESSAGE, message, builder) { httpMessage, metadataProperties ->
+                    httpMessage.addField(METHOD_FIELD, method)
+                    httpMessage.addField(URI_FIELD, uri)
+                    metadataProperties[METHOD_METADATA_PROPERTY] = method
+                    metadataProperties[URI_METADATA_PROPERTY] = uri.toString()
+                }
+
             else -> error("Unsupported message direction: $direction")
         }
 
         return builder.build()
     }
 
+    override fun decode(messageGroup: MessageGroup, context: IReportingContext): MessageGroup {
+        val messages = messageGroup.messages
+
+        require(messages.size == 1) { "Message group must contain only 1 message" }
+
+        val message = messages[0]
+        require(message is RawMessage) { "Message must be a raw message" }
+
+        val body = message.body.toByteArray().toString(UTF_8)
```

Comment:
There is a method `toString(Charset charset)` on `ByteBuf` itself. Maybe we can use that one instead